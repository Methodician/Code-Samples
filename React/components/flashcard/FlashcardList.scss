:local(.FlashcardList) {

  .list-element {
    min-height: 5.5rem;
  }

  .footer-button-container {
    margin-top: 0.5rem;
    width: 100%;
    bottom: 0;
    position: fixed;
    height: 4.5rem;

    .button {
      display: flex;
      align-items: center;
      justify-content: center;
    }

    .flashcards-icon {
      margin-right: 0.8rem;
    }
  }
}
