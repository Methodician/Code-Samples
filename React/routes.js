// This is a good and proper way to manage routing in React
import Explore from 'components/explore/Explore';
import Learn from 'components/learn/Learn';
import Me from 'components/me/Me';
import Review from 'components/review/Review';
import Practice from 'components/practice/Practice';
import Quiz from 'components/quiz/Quiz';
import QuizResults from 'components/quiz/quiz-results/QuizResults';
import Lesson from 'components/lessons/lesson/Lesson';
import FlashcardDeck from 'components/practice/flashcard/flashcard-deck/FlashcardDeck';
import Unit from 'components/unit/Unit.js';
import UnitNav from 'components/learn/unit-nav/UnitNav';

export const routes = {
  EXPLORE: {
    path: '/explore',
    component: Explore,
    linkTo: () => '/explore',
  },
  LEARN: {
    path: '/learn',
    component: Learn,
    linkTo: () => '/learn',
  },
  COURSES: {
    path: '/courses/:courseId',
    component: UnitNav,
    linkTo: courseId => `/courses/${courseId}`,
  },
  ME: {
    path: '/me',
    component: Me,
    linkTo: () => '/me',
  },
  REVIEW: {
    path: '/review',
    component: Review,
    linkTo: () => '/review',
  },
  LESSON: {
    path: '/lessons/:lessonId',
    component: Lesson,
    linkTo: lessonId => `/lessons/${lessonId}`,
  },
  PRACTICE: {
    path: '/lessons/:lessonId/practice',
    component: Practice,
    linkTo: lessonId => `/lessons/${lessonId}/practice`,
  },
  QUIZ: {
    path: '/lessons/:lessonId/quiz',
    component: Quiz,
    linkTo: lessonId => `/lessons/${lessonId}/quiz`,
  },
  QUIZ_RESULTS: {
    path: '/lessons/:lessonId/quiz-results',
    component: QuizResults,
    linkTo: lessonId => `/lessons/${lessonId}/quiz-results`,
  },
  FLASHCARD_DECK: {
    path: '/lessons/:lessonId/flashcard-deck',
    component: FlashcardDeck,
    linkTo: lessonId => `/lessons/${lessonId}/flashcard-deck`,
  },
  UNIT: {
    path: '/units/:unitId',
    component: Unit,
    linkTo: unitId => `/units/${unitId}`,
  },
};
