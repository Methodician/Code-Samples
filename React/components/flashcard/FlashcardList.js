import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import { MdStar as Star } from 'react-icons/md';

import PracticeList from 'components/practice/practice-list/PracticeList';
import SignalButton from 'components/shared/UI/signal-button/SignalButton';
import styles from './FlashcardList.scss';
import PlayAllButton from 'components/shared/UI/play-all-button/PlayAllButton';
import SVGIcon from 'components/shared/UI/icons/SVGIcon';
import { flashcardsD, flashcardsViewBox } from 'components/shared/UI/icons/SVGcoordinates';
import { routes } from 'routes/routes';

import { createFlashcardAudioUrl } from 'services/helpers/contentUrlGenerators';

// REDUX CONNECTIONS
import { connect } from 'react-redux';

import {
  getFlashcardDeck,
  getFlashcardItemIsPlaying,
  getFlashcardAudioSrc,
  getFlashcardListIsPlaying,
  getFlashcardListCurrentIndex,
} from 'selectors/flashcardSelectors';

import { getPracticeAudioSpeed } from 'selectors/practiceSelectors';

import {
  _loadFlashcardDeck,
  _updateFlashcardAudioSrc,
  _updateFlashcardItemIsPlaying,
  _clearFlashcardSession,
  _updateFlashcardListIsPlaying,
  _incrementFlashcardListCurrentIndex,
  _setFlashcardListCurrentIndex,
  _resetFlashcardListIndex,
  _resetFlashcardListIsPlaying,
} from 'actions/flashcardActions';

export class FlashcardList extends PureComponent {
  player = React.createRef();

  static propTypes = {
    loadFlashcardDeck: PropTypes.func.isRequired,
    updateFlashcardAudioSrc: PropTypes.func.isRequired,
    updateFlashcardItemIsPlaying: PropTypes.func.isRequired,
    updateFlashcardListIsPlaying: PropTypes.func.isRequired,
    incrementFlashcardListCurrentIndex: PropTypes.func.isRequired,
    setFlashcardListCurrentIndex: PropTypes.func.isRequired,
    resetFlashcardListIndex: PropTypes.func.isRequired,
    resetFlashcardListIsPlaying: PropTypes.func.isRequired,
    clearFlashcardSession: PropTypes.func.isRequired,
    flashcardDeck: PropTypes.array,
    flashcardItemIsPlaying: PropTypes.bool,
    flashcardListIsPlaying: PropTypes.bool,
    flashcardAudioSrc: PropTypes.string,
    flashcardListCurrentIndex: PropTypes.number,
    practiceAudioSpeed: PropTypes.number,
    history: PropTypes.object.isRequired,
    match: PropTypes.object.isRequired,
  };

  componentDidMount() {
    const { match, loadFlashcardDeck } = this.props;
    const { lessonId } = match.params;
    loadFlashcardDeck(lessonId);
  }

  // add event listeners to audio player for audio play and audio end events
  initPlayer = () => {
    const player = this.player.current;

    player.onended = this.onFlashcardAudioEnded;
    player.onplay = this.onFlashcardAudioPlayed;
  };

  componentDidUpdate(prevProps) {
    // if the player exists and the list has loaded for the first time, add event listeners on the audio element using initPlayer
    if (prevProps.flashcardDeck === null && this.props.flashcardDeck !== null)
      this.initPlayer();

    // if the user clicks play/pause all icon when the the store says the list is not currently playing, play the one item
    if (
      prevProps.flashcardListIsPlaying === false &&
      this.props.flashcardListIsPlaying === true
    ) {
      this.flashcardListPlayItem();
    }

    // if the full list is still playing and the playback of a single item has ended, trigger the playing of the next item in the list
    if (
      prevProps.flashcardItemIsPlaying === true &&
      this.props.flashcardItemIsPlaying === false &&
      this.props.flashcardListIsPlaying === true
    ) {
      this.flashcardListPlayNext();
    }

    // if the user clicks play/pause all icon when the the store says the list is currently playing, pause the item currently being played
    if (
      prevProps.flashcardListIsPlaying === true &&
      this.props.flashcardListIsPlaying === false
    ) {
      this.pauseAudio();
    }

    // if the the user clicks an individual item while play all is happening, the list playback will stop, reset, and item is playing will be set to true, resulting in these conditions being met, and when this happens play the individual list item
    if (
      prevProps.flashcardListIsPlaying === true &&
      this.props.flashcardListIsPlaying === false &&
      this.props.flashcardItemIsPlaying === true
    ) {
      this.handleListItemClicked();
    }

    // if the user changes the audio speed setting while an item is playing, update the audio speed immediately
    if (prevProps.practiceAudioSpeed !== this.props.practiceAudioSpeed) {
      this.player.current.playbackRate = this.props.practiceAudioSpeed;
    }
  }

  componentWillUnmount() {
    const { clearFlashcardSession } = this.props;
    clearFlashcardSession();
  }

  // helper functions

  // general helper function to initiate item playback at the proper speed from the proper audio source
  playAudio = async audioNormal => {
    const player = this.player.current;
    const { updateFlashcardAudioSrc, practiceAudioSpeed } = this.props;

    await updateFlashcardAudioSrc(`${audioNormal}.mp3`);
    player.playbackRate = practiceAudioSpeed;
    player.play();
  };

  // general helper function to pause item playback
  pauseAudio = () => {
    const player = this.player.current;

    player.pause();
  };

  // general helper function to increment the index of the item being played during play all and also to check if the item is the last in the list
  incrementIndex = () => {
    const {
      flashcardDeck,
      flashcardListCurrentIndex,
      incrementFlashcardListCurrentIndex,
    } = this.props;

    incrementFlashcardListCurrentIndex({
      index: flashcardListCurrentIndex,
      deckLength: flashcardDeck.length,
    });
  };

  // methods for playing individual flashcard items on click

  // click event handler for playing individual list item
  handleListItemClicked = async (audioNormal, index) => {
    const player = this.player.current;

    const {
      resetFlashcardListIsPlaying,
      setFlashcardListCurrentIndex,
      flashcardListIsPlaying,
      flashcardItemIsPlaying,
      updateFlashcardItemIsPlaying,
      flashcardAudioSrc,
      flashcardDeck,
    } = this.props;

    // if an individual item is clicked while play all is not active and the item is already playing, pause the item
    if (flashcardListIsPlaying === false && flashcardItemIsPlaying === true) {
      this.pauseAudio();
      updateFlashcardItemIsPlaying(false);
    }
    // if an individual item is clicked while play all is not active and the item is not playing, play the item
    else if (!flashcardListIsPlaying && !flashcardItemIsPlaying) {
      if (
        createFlashcardAudioUrl(flashcardDeck[0].audioSrc) === flashcardAudioSrc
      ) {
        player.currentTime = 0;
        this.playAudio(audioNormal);
      } else {
        setFlashcardListCurrentIndex(index);
        this.playAudio(audioNormal);
        updateFlashcardItemIsPlaying(true);
      }
    }
    // if an individual item is clicked while play all is active, stop play all, reset the item index for play all to 0, and play the item clicked
    else if (flashcardListIsPlaying) {
      // if the item clicked is the same one that is currently playing, reset the audio element's currentTime to 0 before playing
      if (createFlashcardAudioUrl(audioNormal) === flashcardAudioSrc) {
        player.currentTime = 0;
      }

      //then play the item and reset the dialogue list playing trackers to their original state (play from beginning when the button is clicked again)
      this.playAudio(audioNormal);
      resetFlashcardListIsPlaying();
      setFlashcardListCurrentIndex(index);
    }
  };

  onFlashcardAudioEnded = () => {
    const {
      updateFlashcardItemIsPlaying,
      flashcardListIsPlaying,
      resetFlashcardListIndex,
    } = this.props;
    updateFlashcardItemIsPlaying(false);
    // if the list is not playing then set current index back to 0
    if (!flashcardListIsPlaying) resetFlashcardListIndex();
  };

  onFlashcardAudioPlayed = () => {
    const { updateFlashcardItemIsPlaying } = this.props;
    updateFlashcardItemIsPlaying(true);
  };

  // methods for playing and pausing the whole flashcard list

  // click event handler for playing/pausing all
  handlePlayPauseClicked = () => {
    const { flashcardItemIsPlaying, updateFlashcardListIsPlaying } = this.props;

    if (flashcardItemIsPlaying) {
      updateFlashcardListIsPlaying(false);
    } else {
      updateFlashcardListIsPlaying(true);
    }
  };

  // play an individual list item during play all
  flashcardListPlayItem = () => {
    const {
      flashcardDeck,
      flashcardListCurrentIndex,
      resetFlashcardListIndex,
      updateFlashcardListIsPlaying,
    } = this.props;

    if (flashcardListCurrentIndex !== -2) {
      this.playAudio(flashcardDeck[flashcardListCurrentIndex].normal);
    } else {
      resetFlashcardListIndex();
      updateFlashcardListIsPlaying(false);
    }
  };

  // increment the play all item index and then call the method that will play the next item
  flashcardListPlayNext = async () => {
    await this.incrementIndex();
    this.flashcardListPlayItem();
  };

  // method to exit flashcard list and open flashcard deck
  handleStartClick = () => {
    const { lessonId } = this.props.match.params;
    return this.props.history.push(routes.FLASHCARD_DECK.linkTo(lessonId));
  };

  render() {
    const {
      flashcardDeck,
      flashcardAudioSrc,
      flashcardListIsPlaying,
      flashcardItemIsPlaying,
      flashcardListCurrentIndex,
    } = this.props;

    if (!flashcardDeck) return null;

    return (
      <section className={styles.FlashcardList}>
        <PracticeList
          onItemClicked={(audioNormal, index) =>
            this.handleListItemClicked(audioNormal, index)
          }
          practiceMaterials={flashcardDeck}
          activeItemIndex={
            flashcardItemIsPlaying ? flashcardListCurrentIndex : null
          }
        >
          <Star className="icon star"/>
        </PracticeList>
        <audio
          id="flashcards-list-audio"
          src={flashcardAudioSrc}
          ref={this.player}
        />
        <footer className="footer-button-container">
          <SignalButton
            buttonType="neutral"
            click={this.handleStartClick}
          >
            <>
              <SVGIcon
                width={30}
                height={24}
                className="flashcards-icon"
                viewBox={flashcardsViewBox}
                d={flashcardsD}
              />
              <span className="button-text">
                Practice
              </span>
            </>
          </SignalButton>
          <PlayAllButton
            click={this.handlePlayPauseClicked}
            isListPlaying={flashcardListIsPlaying}
          />
        </footer>
      </section>
    );
  }
}

const mapStateToProps = state => ({
  flashcardDeck: getFlashcardDeck(state),
  flashcardItemIsPlaying: getFlashcardItemIsPlaying(state),
  flashcardAudioSrc: getFlashcardAudioSrc(state),
  practiceAudioSpeed: getPracticeAudioSpeed(state),
  flashcardListIsPlaying: getFlashcardListIsPlaying(state),
  flashcardListCurrentIndex: getFlashcardListCurrentIndex(state),
});

const mapDispatchToProps = dispatch => ({
  loadFlashcardDeck: lessonId => dispatch(_loadFlashcardDeck(lessonId)),
  updateFlashcardAudioSrc: listItemId =>
    dispatch(_updateFlashcardAudioSrc(listItemId)),
  updateFlashcardItemIsPlaying: flashcardItemIsPlaying =>
    dispatch(_updateFlashcardItemIsPlaying(flashcardItemIsPlaying)),
  updateFlashcardListIsPlaying: flashcardListIsPlaying =>
    dispatch(_updateFlashcardListIsPlaying(flashcardListIsPlaying)),
  clearFlashcardSession: () => dispatch(_clearFlashcardSession()),
  incrementFlashcardListCurrentIndex: indexAndLength =>
    dispatch(_incrementFlashcardListCurrentIndex(indexAndLength)),
  setFlashcardListCurrentIndex: index =>
    dispatch(_setFlashcardListCurrentIndex(index)),
  resetFlashcardListIndex: () => dispatch(_resetFlashcardListIndex()),
  resetFlashcardListIsPlaying: () => dispatch(_resetFlashcardListIsPlaying()),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(FlashcardList);
