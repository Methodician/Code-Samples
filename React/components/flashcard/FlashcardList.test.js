import React from 'react';
import { shallow } from 'enzyme';
import { FlashcardList } from './FlashcardList';

const mockFn = jest.fn();
const flashcardDeck = [
  {
    id: 'BCC-002-01-001',
    cardType: 'B',
    pinyin: 'yī',
    characters: '一',
    english: 'one',
    audioNormal: 'BCC-002-01-001-N',
    audioSlow: 'BCC-002-01-002-S',
  },
  {
    id: 'BCC-002-01-002',
    cardType: 'B',
    pinyin: 'èr',
    characters: '二',
    english: 'two',
    audioNormal: 'BCC-002-01-002-N',
    audioSlow: 'BCC-002-01-003-S',
  },
  {
    id: 'BCC-002-01-003',
    cardType: 'B',
    pinyin: 'sān',
    characters: '三',
    english: 'three',
    audioNormal: 'BCC-002-01-003-N',
    audioSlow: 'BCC-002-01-004-S',
  },
  {
    id: 'BCC-002-01-004',
    cardType: 'B',
    pinyin: 'sì',
    characters: '四',
    english: 'four',
    audioNormal: 'BCC-002-01-004-N',
    audioSlow: 'BCC-002-01-005-S',
  },
  {
    id: 'BCC-002-01-005',
    cardType: 'B',
    pinyin: 'wǔ',
    characters: '五',
    english: 'five',
    audioNormal: 'BCC-002-01-005-N',
    audioSlow: 'BCC-002-01-006-S',
  },
  {
    id: 'BCC-002-01-006',
    cardType: 'B',
    pinyin: 'liù',
    characters: '六',
    english: 'six',
    audioNormal: 'BCC-002-01-006-N',
    audioSlow: 'BCC-002-01-007-S',
  },
  {
    id: 'BCC-002-01-007',
    cardType: 'B',
    pinyin: 'qī',
    characters: '七',
    english: 'seven',
    audioNormal: 'BCC-002-01-007-N',
    audioSlow: 'BCC-002-01-008-S',
  },
  {
    id: 'BCC-002-01-008',
    cardType: 'B',
    pinyin: 'bā',
    characters: '八',
    english: 'eight',
    audioNormal: 'BCC-002-01-008-N',
    audioSlow: 'BCC-002-01-009-S',
  },
  {
    id: 'BCC-002-01-009',
    cardType: 'B',
    pinyin: 'jiǔ',
    characters: '九',
    english: 'nine',
    audioNormal: 'BCC-002-01-009-N',
    audioSlow: 'BCC-002-01-010-S',
  },
  {
    id: 'BCC-002-01-010',
    cardType: 'B',
    pinyin: 'shí',
    characters: '十',
    english: 'ten',
    audioNormal: 'BCC-002-01-010-N',
    audioSlow: 'BCC-002-01-011-S',
  },
  {
    id: 'BCC-002-01-011',
    cardType: 'B',
    pinyin: 'líng',
    characters: '零',
    english: 'zero',
    audioNormal: 'BCC-002-01-011-N',
    audioSlow: 'BCC-002-01-001-S',
  },
];

const firstCardAudioSource =
  'https://cdn.yoyochinese.com/audio/practice/BCC-002-01-001-N.mp3';

const mockMatch = { params: { lessonId: '06' } };

const comp = (
  <FlashcardList
    loadFlashcardDeck={mockFn}
    updateFlashcardAudioSrc={mockFn}
    updateFlashcardItemIsPlaying={mockFn}
    updateFlashcardListIsPlaying={mockFn}
    incrementFlashcardListCurrentIndex={mockFn}
    setFlashcardListCurrentIndex={mockFn}
    resetFlashcardListIndex={mockFn}
    resetFlashcardListIsPlaying={mockFn}
    clearFlashcardSession={mockFn}
    history={{}}
    match={mockMatch}
    flashcardDeck={flashcardDeck}
    flashcardAudioSrc={firstCardAudioSource}
  />
);
const shallowComp = shallow(comp);

describe('FlashcardList Component', () => {
  it('should match a snapshot', () => {
    expect(shallowComp).toMatchSnapshot();
  });

  it('should render audio with given source', () => {
    expect(shallowComp.find('audio').html()).toContain(firstCardAudioSource);
  });

  describe('Footer tests:', () => {
    const footer = shallowComp.find('footer');

    it('should render a footer with the expected class', () => {
      expect(footer.hasClass('footer-button-container')).toEqual(true);
    });

    it('should contain a neutral signal button', () => {
      expect(footer.find('SignalButton').html()).toContain('button neutral');
    });

    it('should contain a PlayAllButton', () => {
      expect(footer.find('PlayAllButton').html()).toContain('button play-icon');
    });
  });
});
