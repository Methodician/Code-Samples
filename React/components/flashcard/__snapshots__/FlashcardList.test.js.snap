// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlashcardList Component should match a snapshot 1`] = `
<section
  className="FlashcardList"
>
  <Connect(PracticeList)
    activeItemIndex={null}
    onItemClicked={[Function]}
    practiceMaterials={
      Array [
        Object {
          "audioNormal": "BCC-002-01-001-N",
          "audioSlow": "BCC-002-01-002-S",
          "cardType": "B",
          "characters": "一",
          "english": "one",
          "id": "BCC-002-01-001",
          "pinyin": "yī",
        },
        Object {
          "audioNormal": "BCC-002-01-002-N",
          "audioSlow": "BCC-002-01-003-S",
          "cardType": "B",
          "characters": "二",
          "english": "two",
          "id": "BCC-002-01-002",
          "pinyin": "èr",
        },
        Object {
          "audioNormal": "BCC-002-01-003-N",
          "audioSlow": "BCC-002-01-004-S",
          "cardType": "B",
          "characters": "三",
          "english": "three",
          "id": "BCC-002-01-003",
          "pinyin": "sān",
        },
        Object {
          "audioNormal": "BCC-002-01-004-N",
          "audioSlow": "BCC-002-01-005-S",
          "cardType": "B",
          "characters": "四",
          "english": "four",
          "id": "BCC-002-01-004",
          "pinyin": "sì",
        },
        Object {
          "audioNormal": "BCC-002-01-005-N",
          "audioSlow": "BCC-002-01-006-S",
          "cardType": "B",
          "characters": "五",
          "english": "five",
          "id": "BCC-002-01-005",
          "pinyin": "wǔ",
        },
        Object {
          "audioNormal": "BCC-002-01-006-N",
          "audioSlow": "BCC-002-01-007-S",
          "cardType": "B",
          "characters": "六",
          "english": "six",
          "id": "BCC-002-01-006",
          "pinyin": "liù",
        },
        Object {
          "audioNormal": "BCC-002-01-007-N",
          "audioSlow": "BCC-002-01-008-S",
          "cardType": "B",
          "characters": "七",
          "english": "seven",
          "id": "BCC-002-01-007",
          "pinyin": "qī",
        },
        Object {
          "audioNormal": "BCC-002-01-008-N",
          "audioSlow": "BCC-002-01-009-S",
          "cardType": "B",
          "characters": "八",
          "english": "eight",
          "id": "BCC-002-01-008",
          "pinyin": "bā",
        },
        Object {
          "audioNormal": "BCC-002-01-009-N",
          "audioSlow": "BCC-002-01-010-S",
          "cardType": "B",
          "characters": "九",
          "english": "nine",
          "id": "BCC-002-01-009",
          "pinyin": "jiǔ",
        },
        Object {
          "audioNormal": "BCC-002-01-010-N",
          "audioSlow": "BCC-002-01-011-S",
          "cardType": "B",
          "characters": "十",
          "english": "ten",
          "id": "BCC-002-01-010",
          "pinyin": "shí",
        },
        Object {
          "audioNormal": "BCC-002-01-011-N",
          "audioSlow": "BCC-002-01-001-S",
          "cardType": "B",
          "characters": "零",
          "english": "zero",
          "id": "BCC-002-01-011",
          "pinyin": "líng",
        },
      ]
    }
  >
    <MdStar
      className="icon star"
    />
  </Connect(PracticeList)>
  <audio
    id="flashcards-list-audio"
    src="https://cdn.yoyochinese.com/audio/practice/BCC-002-01-001-N.mp3"
  />
  <footer
    className="footer-button-container"
  >
    <SignalButton
      buttonType="neutral"
      click={[Function]}
    >
      <SVGIcon
        className="flashcards-icon"
        d="M27.55,2.17L17.12,0a1.83,1.83,0,0,0-2,1H8.55A1.8,1.8,0,0,0,6.73,2.76V2.82L1.44,3.9A1.76,1.76,0,0,0,0,5.95V6L3.42,21.6A1.81,1.81,0,0,0,5.57,23L17,20.63a1.88,1.88,0,0,0,.46-0.17h2.27l3.69,0.76a1.81,1.81,0,0,0,2.15-1.36L29,4.24a1.76,1.76,0,0,0-1.37-2.07h0Zm-22,18.61a0.2,0.2,0,0,1-.22-0.15L2.16,6.13a0.18,0.18,0,0,1,.15-0.22L6.73,5v13.7a1.75,1.75,0,0,0,1.07,1.6ZM9,18.32a0.18,0.18,0,0,1-.19-0.18h0V3.33A0.2,0.2,0,0,1,9,3.14H14.6l-0.8,3.7L11.6,17a1.7,1.7,0,0,0,.25,1.3H9ZM23.7,18.89a0.2,0.2,0,0,1-.22.15l-1.41-.3-2.19-.44L18.3,18l-2.19-.44-2.25-.46a0.18,0.18,0,0,1-.15-0.21h0l1.13-5.2,1-4.83,0.8-3.7,0.17-.77a0.2,0.2,0,0,1,.22-0.15l5,1,4.63,1a0.18,0.18,0,0,1,.15.21h0Zm-2-12.59,0.87,0.24c0.27,0.08.32,0.23,0.16,0.45L21.15,9S21,9.14,21.07,9.2a0.18,0.18,0,0,0,.16,0l2,0.07h0.15a0.28,0.28,0,0,1,.25.16,0.23,0.23,0,0,1-.07.27L22.28,11.2,18.8,15.08l-0.07.08a0.29,0.29,0,0,1-.4.07,0.24,0.24,0,0,1-.07-0.33l0,0,1.3-2.2,1-1.73c0.07-.11.06-0.14-0.08-0.14l-2.15-.07H18.22A0.25,0.25,0,0,1,18,10.3l1.64-2.79c0.25-.43.51-0.85,0.75-1.28a0.32,0.32,0,0,1,.41-0.17Z"
        height={24}
        viewBox="0 0 30 24"
        width={30}
      />
      <span
        className="button-text"
      >
        Practice
      </span>
    </SignalButton>
    <PlayAllButton
      click={[Function]}
    />
  </footer>
</section>
`;
